Google.Cloud.Spanner.NHibernate/ICriteriaExtensions.cs

```diff
@@ -98,7 +98,127 @@ public static ICriteria SetStatementAndTableHints(this ICriteria query, string s
+        /// </param>
+        /// <param name="hint">The hint to add to the join, e.g. @{JOIN_METHOD=HASH_JOIN}</param>
+        /// <returns>The query including the join hint</returns>
+        public static ICriteria SetJoinHint(this ICriteria query, string table, string hint)
+        {
+            GaxPreconditions.CheckNotNull(query, nameof(query));
+            return query.SetComment($"{SpannerQueryHintInterceptor.SpannerQueryHintsPrefix}{SpannerQueryHintInterceptor.SpannerJoinHintPrefix}`{table}`{hint}");
+        }
+        
+        /// <summary>
+        /// Sets the join hints to use for the given query.
+        /// See https://cloud.google.com/spanner/docs/reference/standard-sql/query-syntax#join_hints for more
+        /// information on join hints.
+        /// 
+        /// Calling this method will override any other hint(s) that may already have been set on the query.
+        /// </summary>
+        /// <param name="query">The query to set the join hint on</param>
+        /// <param name="hints">
+        /// A dictionary containing the hints to use. The key of each entry should be the unquoted name of the
+        /// right-hand table of the join.
+        /// Example: `SELECT * FROM Albums LEFT JOIN Singers ...` requires the table name Singers to be specified to add
+        /// a hint to the join.
+        /// The value should be the hint to add to the join, e.g. @{JOIN_METHOD=HASH_JOIN}
+        /// </param>
+        /// <returns>The query including the join hints</returns>
+        public static ICriteria SetJoinHints(this ICriteria query, Dictionary<string, string> hints)
+        {
+            GaxPreconditions.CheckNotNull(query, nameof(query));
+            GaxPreconditions.CheckNotNull(hints, nameof(hints));
+            var builder = new StringBuilder($"{SpannerQueryHintInterceptor.SpannerQueryHintsPrefix}");
+            return query.SetComment(SpannerQueryHintInterceptor.AppendJoinHintsString(builder, hints).ToString());
+        }
+
+        /// <summary>
+        /// Sets the statement and join hints to use for the given query.
+        /// Calling this method will override any other hint(s) that may already have been set on the query.
+        /// </summary>
+        /// <param name="query">The query to set the statement and join hints on</param>
+        /// <param name="statementHint">The statement hint to use for the query, e.g. @{OPTIMIZER_VERSION=1}</param>
+        /// <param name="joinHints">
+        /// A dictionary containing the hints to use. The key of each entry should be the unquoted name of the
+        /// right-hand table of the join.
+        /// The value should be the hint to add to the table, e.g. @{JOIN_METHOD=HASH_JOIN}
+        /// </param>
+        /// <returns>The query including the statement and join hints</returns>
+        public static ICriteria SetStatementAndJoinHints(this ICriteria query, string statementHint, Dictionary<string, string> joinHints)
```

Comment:
Instead of having these specific combinations, can we do something like the following (as you would in Java and even in the ADO.net client)?

```
session
.Query<Album>()
.Select(a => a.Singer)
.SetJoinHint(<hint>)
.SetStatementHints(<hint>)
.SetTableHint(<hint>)
...
```

Reply:
That is unfortunately difficult, as:
1. `ICriteria` (and the other interfaces that define queries in NHibernate) do not expose an API like `AddComment`, only `SetComment`.
2. `ICriteria` also does not have any other data structure where we could (temporarily) store one or more hints before the query string is generated.
3. `ICriteria` does not have an API that allows us to read back any comments that have already been set, so we cannot do something like `query.SetComment(query.GetComment() + " ... ")`

What could be possible is to create a `HintsBuilder` class with such methods, and then an extension method like `ICriteria.SetHints(Hints hints)`. So the code would then look like something like this:


```csharp
session
  .Query<Album>()
  .Select(a => a.Singer)
  .SetHints(Hints
    .NewBuilder()
    .SetJoinHint(<hint>)
    .SetStatementHints(<hint>)
    .SetTableHint(<hint>)
    .Build())
...

```

Reply:
I've refactored the hint code so it uses a `Hints` class that can contain multiple different types of hints and added a builder class for it. That should make it easier for users to create queries with multiple different hints.

Reply:
Awesome, thanks so much for doing this. I think it's much cleaner now.